blockchain/blockchain_test.go

```diff
@@ -1644,3 +1658,403 @@ func TestBlockchain_WriteFullBlock(t *testing.T) {
 	require.NotNil(t, db[hex.EncodeToHex(getKey(storage.RECEIPTS, header.Hash.Bytes()))])
 }
+
+func TestDiskUsageWriteBatchAndUpdate(t *testing.T) {
+	const (
+		checkInterval  = 100 * time.Millisecond
+		numberOfBlocks = 100
+		blockTime      = 1 * time.Nanosecond
+	)
+
+	blockJSONFile, err := os.Open("testfiles/testblock.json")
+	require.NoError(t, err)
+
+	blockBytes, err := io.ReadAll(blockJSONFile)
+	require.NoError(t, err)
+
+	receiptsJSONFile, err := os.Open("testfiles/testreceipts.json")
+	require.NoError(t, err)
+
+	receiptsBytes, err := io.ReadAll(receiptsJSONFile)
+	require.NoError(t, err)
+
+	blockWriter(t, numberOfBlocks, blockTime, checkInterval, blockBytes, receiptsBytes)
+}
+
+func blockWriter(tb testing.TB, numberOfBlocks uint64, blockTime, checkInterval time.Duration, byteToRead []byte, receiptsBytesToRead []byte) {
+	tb.Helper()
+
+	blockTicker := time.NewTicker(blockTime)
+
+	type blockCounter struct {
+		mu sync.RWMutex
+		x  uint64
+	}
+
+	counter := &blockCounter{x: 0}
+
+	quitChan := make(chan bool)
+
+	p := fmt.Sprintf("/tmp/blockchain-disk-usage-test")
+	err := os.Mkdir(p, 0775)
+	require.NoError(tb, err)
+
+	tb.Cleanup(func() {
+		if err := os.RemoveAll(p); err != nil {
+			tb.Fatal(err)
+		}
+	})
+
+	db, err := leveldb.NewLevelDBStorage(
+		filepath.Join(p),
+		hclog.NewNullLogger(),
+	)
+	require.NoError(tb, err)
+
+	block, err := customJSONBlockUnmarshall(tb, byteToRead)
+	require.NoError(tb, err)
+
+	receipts, err := customJSONReceiptsUnmarshall(tb, receiptsBytesToRead)
+	require.NoError(tb, err)
+
+	dirSizeCheck := func() {
+		ticker := time.NewTicker(checkInterval)
+
+		for {
+			select {
+			case <-quitChan:
+				return
+			case <-ticker.C:
+				dirSizeValue, err := dirSize(tb, p)
+
+				if err != nil {
+					tb.Log(err)
+				}
+
+				counter.mu.RLock()
+
+				tb.Logf("BLOCK: %d DIRSIZE IS: %d and average is:%.2f", counter.x, dirSizeValue, float64(dirSizeValue)/float64(counter.x))
+
+				counter.mu.RUnlock()
+			}
+		}
+	}
+
+	blockchain := &Blockchain{db: db}
+
+	dirSizeBeforeBlocks, err := dirSize(tb, p)
+	require.NoError(tb, err)
+	tb.Logf("DIRSIZE IS: %d", dirSizeBeforeBlocks)
+
+	go dirSizeCheck()
+
+	for ; ; <-blockTicker.C {
+		batchWriter := storage.NewBatchWriter(db)
+
+		counter.mu.RLock()
+		block.Block.Header.Number = counter.x
+		block.Block.Header.Hash = types.StringToHash(fmt.Sprintf("%d", counter.x))
+
+		for i, transaction := range block.Block.Transactions {
+			transaction.Nonce = counter.x * uint64(i)
+			addr := types.StringToAddress(fmt.Sprintf("%d", counter.x*uint64(i)))
+			transaction.To = &addr
+		}
+
+		counter.mu.RUnlock()
+
+		batchWriter.PutHeader(block.Block.Header)
+		batchWriter.PutBody(block.Block.Hash(), block.Block.Body())
+
+		batchWriter.PutReceipts(block.Block.Hash(), receipts)
+
+		require.NoError(tb, blockchain.writeBatchAndUpdate(batchWriter, block.Block.Header, big.NewInt(0), false))
+
+		counter.mu.Lock()
+
+		counter.x++
+
+		counter.mu.Unlock()
```

Comment:
It would be probably a bit nicer if we would have getter and setter provided for `blockCounter` struct (especially would be convenient because of releasing read lock when reading the value down there in the if statement, because of the `break` statement and invoking `RUnlock` twice)

```go
if counter.x == numberOfBlocks {
			counter.mu.RUnlock()

			break
		}

		counter.mu.RUnlock()
```